frontend/src/lib/HttpClient.ts

```diff
@@ -0,0 +1,56 @@
+/**
+ * @license
+ * Copyright 2018-2020 Streamlit Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import axiosLib, { AxiosInstance } from "axios"
+import { BaseUriParts } from "lib/UriUtil"
+import { getCookie } from "lib/utils"
+
+/**
+ * Base class for HTTP Clients
+ */
+export default class HttpClient {
+  protected readonly getServerUri: () => BaseUriParts | undefined
+  protected axiosInstance: AxiosInstance
+  protected readonly csrfEnabled: boolean
+
+  public constructor(getServerUri: () => BaseUriParts | undefined) {
+    const xsrf_cookie = getCookie("_xsrf")
+    this.getServerUri = getServerUri
+    this.csrfEnabled = !!xsrf_cookie
+    this.axiosInstance = axiosLib.create(
+      this.csrfEnabled
+        ? {
+            withCredentials: true,
+            headers: {
+              "X-Xsrftoken": xsrf_cookie,
+            },
+          }
+        : {}
+    )
+  }
+
+  public updateCsrfToken(): void {
+    const csrfCookie = getCookie("_xsrf")
+    if (csrfCookie) {
+      this.axiosInstance.defaults.headers["X-Xsrftoken"] = csrfCookie
+      this.axiosInstance.defaults.withCredentials = true
+    } else {
+      delete this.axiosInstance.defaults.headers["X-Xsrftoken"]
+      this.axiosInstance.defaults.withCredentials = false
+    }
```

Comment:
I debated about this for quite a bit and ended up going with this approach to provide more flexibility.

On the server side we have to update the allowable headers for each request handler so its possible we could have CSRF enabled on some endpoints and not others (not sure why we would 🤷‍♀️) I'm assuming we'll end up mapping a `HTTPClient` to a `tornado.RequestHandler`. In that regards, if we create a new RequestHandler without specifying the necessary allowable headers, the `HTTPClient` will not cry since I assume we wouldn't have called `updateCsrfToken` for the new `HTTPClient` as well. This could be a completely moot use case I'm accommodating for. Let me know what you think and happy to change!

Reply:
I don't quite follow, but I still think that being stateless is accomplishable and desirable. (Though again, I might be misunderstanding something!) 

If the CSRF header is a per-client concern, then we could have "useCSRFToken" as a parameter that gets passed to the constructor, and the code above becomes:

```javascript
// (in class HTTPClient):
public request(params: AxiosRequestConfig) {
  if (this.useCSRFToken) {
      const cookie = getCookie("_xsrf")
      if (cookie != null) {
        params.headers["X-Xsrftoken"] = cookie
        params.withCredentials = true
      }
  }
  return axios.request(params)
}
```

(It's the stateful `updateCsrfToken()` function that I'm objecting to.)

Reply:
I didn't think about passing it at the constructor level. I only thought about passing it at the request level and was too saddened by that. I'll update!